File: db4o.net/Db4oUnit.Extensions/Db4oUnit.Extensions/FieldIndexAssert.cs
/* Copyright (C) 2004 - 2011  Versant Inc.  http://www.db4o.com */

using System;
using Db4oUnit;
using Db4objects.Db4o.Foundation;
using Db4objects.Db4o.Internal;
using Db4objects.Db4o.Internal.Btree;

namespace Db4oUnit.Extensions
{
	public class FieldIndexAssert
	{
		private readonly Type _clazz;

		private readonly string _name;

		public FieldIndexAssert(Type clazz, string name)
		{
			_clazz = clazz;
			_name = name;
		}

		public virtual void AssertSingleEntry(LocalObjectContainer container, long id)
		{
			BooleanByRef called = new BooleanByRef();
			Index(container).TraverseKeys(container.SystemTransaction(), new _IVisitor4_24(id
				, called));
			Assert.IsTrue(called.value);
		}

		private sealed class _IVisitor4_24 : IVisitor4
		{
			public _IVisitor4_24(long id, BooleanByRef called)
			{
				this.id = id;
				this.called = called;
			}

			public void Visit(object key)
			{
				Assert.AreEqual(id, ((IFieldIndexKey)key).ParentID());
				Assert.IsFalse(called.value);
				called.value = true;
			}

			private readonly long id;

			private readonly BooleanByRef called;
		}

		private BTree Index(LocalObjectContainer container)
		{
			return FieldMetadata(container).GetIndex(null);
		}

		private Db4objects.Db4o.Internal.FieldMetadata FieldMetadata(LocalObjectContainer
			 container)
		{
			return ClassMetadata(container).FieldMetadataForName(_name);
		}

		private Db4objects.Db4o.Internal.ClassMetadata ClassMetadata(LocalObjectContainer
			 container)
		{
			return container.ClassMetadataForReflectClass(container.Reflector().ForClass(_clazz
				));
		}
	}
}

File: db4ounit.extensions/src/db4ounit/extensions/FieldIndexAssert.java
/* Copyright (C) 2009  Versant Inc.  http://www.db4o.com */

package db4ounit.extensions;

import com.db4o.foundation.*;
import com.db4o.internal.*;
import com.db4o.internal.btree.*;

import db4ounit.*;

public class FieldIndexAssert {
	
	private final Class _clazz;
	
	private final String _name;
	
	public FieldIndexAssert(Class clazz, String name){
		_clazz = clazz;
		_name = name;
	}
	
	public void assertSingleEntry(LocalObjectContainer container, final long id) {
		final BooleanByRef called = new BooleanByRef();
		index(container).traverseKeys(container.systemTransaction(), new Visitor4<FieldIndexKey>() {
			public void visit(FieldIndexKey key) {
				Assert.areEqual(id, key.parentID());
				Assert.isFalse(called.value);
				called.value = true;
			}
		});
		Assert.isTrue(called.value);
	}

	private BTree index(LocalObjectContainer container) {
		return fieldMetadata(container).getIndex(null);
	}

	private FieldMetadata fieldMetadata(LocalObjectContainer container) {
		return classMetadata(container).fieldMetadataForName(_name);
	}

	private ClassMetadata classMetadata(LocalObjectContainer container) {
		return container.classMetadataForReflectClass(container.reflector().forClass(_clazz));
	}


}
